### PR TITLE
fix: directory module name checking

### PR DESCRIPTION
**Pull request by Alizter**
fix https://github.com/ocaml/dune/issues/7600

### CODE REVIEW

File: src/dune_rules/ml_sources.ml [hunk 1]
```diff
@@ -456,7 +456,10 @@ let make dune_file ~dir ~scope ~lib_config ~loc ~lookup_vlib
       | Include Qualified ->
         List.fold_left dirs ~init:Module_trie.empty
           ~f:(fun acc ((dir : Path.Build.t), local, files) ->
-            let path = List.map local ~f:Module_name.of_string in
+            let path =
+              List.map local ~f:(fun m ->
+                  Module_name.parse_string_exn (Loc.none, m))
```
Comment (rgrinberg) [inline src/dune_rules/ml_sources.ml:461]: Can you set the location here using `Loc_in_dir` here?

**Review by anmonteiro — COMMENTED** [review]
Could we mangle invalid names instead of erroring?

**Comment by Alizter** [comment]
@anmonteiro If we mangle, how is the user supposed to refer to it in their file?

**Comment by anmonteiro** [comment]
I think it'd be beneficial to support some of those directories and document / log the mangling scheme. For melange specifically, we'll probably want to have directories called `[id]` and `(layout)` (check https://beta.nextjs.org/docs/routing/defining-routes).

**Comment by Alizter** [comment]
I suppose another solution would be to make the module name of a directory configurable from a stanza. So we could do
```
(subdir foo-bar (qualified_name foo_bar))
```
That way when Dune wants to give module names corresponding to directories, it would query a naming map which would be overridden with these stanzas.

**Comment by anmonteiro** [comment]
I like a solution along those lines. We could even reuse `include_subdirs` for it:


```
(include_subdirs 
  (mode qualified)
  ((src foo) (src/invalid-[module]-name invalid_module_name)))
```

**Comment by anmonteiro** [comment]
(And it goes without saying: I don't think these should be implemented in this PR). The bug fix is sensible and looks good to me.

**Comment by Alizter** [comment]
@anmonteiro Can you open a ticket with this feature request then?

**Comment by anmonteiro** [comment]
https://github.com/ocaml/dune/issues/7605

**Review by rgrinberg — APPROVED** [review]

**Review by rgrinberg — APPROVED** [review]

**Comment by Niols** [comment]
I think the feature request in #7605 makes a lot of sense. I guess two other solutions to a directory with a malformed module name would be:

1. The directory in question contains only data and one uses `(data_only_dirs bar-baz)` (or `data_only_dirs` on an ancestor of the directory with a malformed module name).

2. The directory in question could be ignored altogether. (Is there a stanza for this?)

Should I open feature requests for those as well?

**Comment by Alizter** [comment]
> 1. The directory in question contains only data and one uses `(data_only_dirs bar-baz)` (or `data_only_dirs` on an ancestor of the directory with a malformed module name).

I am not sure what exactly you mean here. Could you open an issue explaining this further?

> 2. The directory in question could be ignored altogether. (Is there a stanza for this?)

You can use the `(dirs)` stanza to remove a directory from Dunes vision. Something like `(dirs :standard \ ignored_dir)`.

**Comment by Niols** [comment]
cf https://github.com/ocaml/dune/pull/7620 for two examples. One is your suggestion for `(dirs)`, which I didn't know about, and which works like a charm. The other one uses `(data_only_dirs)`, hoping that Dune will be fine with a directory then having an invalid module name. This is however not the case. I will open a feature request for this.

**Comment by Niols** [comment]
Feature request in question: https://github.com/ocaml/dune/issues/7628.

**Comment by rgrinberg** [comment]
Yes

**Pull request merged**